File: LeetCode-Java/Binary Tree Zigzag Level Order Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        LinkedList<TreeNode> nodeList = new LinkedList<TreeNode>();
        int turn = 0;
        int length = 1;
        List<List<Integer>> result = new ArrayList<List<Integer>>();
        if(root == null) 
            return result;
        nodeList.addLast(root);
        while(length > 0)
        {
            int newLength = 0;
            List<Integer> tempResult = new ArrayList<Integer>();
            while(length > 0)
            {
                if(turn == 0)
                {
                    TreeNode tempNode = nodeList.pollFirst();
                    tempResult.add(tempNode.val);
                    if(tempNode.left != null)
                    {
                        nodeList.addLast(tempNode.left);
                        newLength ++;
                    }
                    if(tempNode.right != null)
                    {
                        nodeList.addLast(tempNode.right);
                        newLength ++;
                    }
                }
                else
                {
                    TreeNode tempNode = nodeList.pollLast();
                    tempResult.add(tempNode.val);
                    if(tempNode.right != null)
                    {
                        nodeList.addFirst(tempNode.right);
                        newLength ++;
                    }
                    if(tempNode.left != null)
                    {
                        nodeList.addFirst(tempNode.left);
                        newLength ++;
                    }
                }
                length --;
            }
            turn = 1 - turn;
            length = newLength;
            result.add(tempResult);
        }
        return result;
    }
}

File: Leetcode-cpp/binary_tree_zig_zag_level_order_traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> result;
        list<TreeNode*> record;
        if (root == nullptr) return result;
        int count = 1;
        record.push_back(root);
        bool is_forward = false;
        while (!record.empty()) {
            is_forward = !is_forward;
            vector<int> temp_result;
            int new_count = 0;
            while (count > 0) {
                if (is_forward) {
                    TreeNode* current = record.front();
                    record.pop_front();
                    count--;
                    temp_result.push_back(current->val);
                    if (current->left != nullptr) {
                        record.push_back(current->left);
                        new_count++;
                    }
                    if (current->right != nullptr) {
                        record.push_back(current->right);
                        new_count++;
                    }
                } else {
                    TreeNode* current = record.back();
                    record.pop_back();
                    count--;
                    temp_result.push_back(current->val);
                    if (current->right != nullptr) {
                        record.push_front(current->right);
                        new_count++;
                    }
                    if (current->left != nullptr) {
                        record.push_front(current->left);
                        new_count++;
                    }
                }
            }
            result.push_back(temp_result);
            count = new_count;
        }
        return result;
    }
};